Timeline.cs

```diff
@@ -0,0 +1,506 @@
+                        select pair).ToList();
+        }
+        return result;
+    }
+
+    /// <summary>
+    /// Convert the timeline to an array.
+    /// </summary>
+    internal (DateTime Time, TValue Value)[] ToArray()
+    {
+        return timeline.ToArray();
+    }
+
+    /// <summary>
+    /// Convert the timeline to a list.
+    /// </summary>
+    internal IList<(DateTime Time, TValue Value)> ToList()
+    {
+        return timeline;
+    }
+
+    /// <summary>
+    /// Convert the timeline to a dictionary.
+    /// </summary>
+    internal IDictionary<DateTime, TValue> ToDictionary()
+    {
+        var dictionary = new Dictionary<DateTime, TValue>();
+        foreach (var (date, time) in timeline)
+            dictionary.Add(date, time);
+        return dictionary;
+    }
+
+    /// <summary>
+    /// Returns true if <paramref name="obj"/> is equal to this timeline.
+    /// </summary>
+    /// <param name="obj">A <see cref="Timeline{TValue}"/> instance.</param>
+    public override bool Equals(object obj)
+    {
+        return this == obj as Timeline<TValue>;
+    }
+
+    public override int GetHashCode()
+    {
+        return base.GetHashCode();
+    }
+
+    void ICollection<(DateTime Time, TValue Value)>.Add((DateTime Time, TValue Value) item)
+    {
+        Add(item.Time, item.Value);
+    }
+
+    bool ICollection<(DateTime Time, TValue Value)>.Contains((DateTime Time, TValue Value) item)
+    {
+        return Contains(item.Time, item.Value);
+    }
+
+    bool ICollection<(DateTime Time, TValue Value)>.Remove((DateTime Time, TValue Value) item)
+    {
+        return Remove(item.Time, item.Value);
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return new TimelineEnumerator(this);
+    }
+
+    IEnumerator<(DateTime Time, TValue Value)> IEnumerable<(DateTime Time, TValue Value)>.GetEnumerator()
+    {
+        return new TimelineEnumerator(this);
+    }
+
+    /// <summary>
+    /// Returns true if <paramref name="left"/> is equal to <paramref name="right"/>.
+    /// </summary>
+    public static bool operator ==(Timeline<TValue> left, Timeline<TValue> right)
+    {
+        var leftArray = left.ToArray();
+        var rightArray = right.ToArray();
+        if (leftArray.Length == rightArray.Length)
+        {
+            for (int i = 0; i < leftArray.Length; i++)
+            {
+                if (leftArray[i].Time != rightArray[i].Time && !leftArray[i].Value.Equals(rightArray[i].Value))
+                    return false;
+            }
+            return true;
+        }
+        else
+            return false;
+    }
+
+    /// <summary>
+    /// Returns true if <paramref name="left"/> is unequal to <paramref name="right"/>.
+    /// </summary>
+    public static bool operator !=(Timeline<TValue> left, Timeline<TValue> right)
+    {
+        return !(left == right);
+    }
+
+    private class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>, IDisposable
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)